src/traces/scatter/plot.js

```diff
@@ -458,8 +458,10 @@ function plotOne(gd, idx, plotinfo, cdscatter, cdscatterAll, element, transition
         join.enter().append('g').classed('textpoint', true).append('text');
 
         join.each(function(d) {
-            var sel = transition(d3.select(this).select('text'));
-            Drawing.translatePoint(d, sel, xa, ya);
+            var g = d3.select(this);
+            var sel = transition(g.select('text'));
+            var hasTextPt = Drawing.translatePoint(d, sel, xa, ya);
+            if(!hasTextPt) g.remove();
```

Comment:
Each `<text>` node is wrapped in a `<g class="textpoint">` group. It is on this group that the data-join happens so we must make sure that group is removed whenever `<text>` node are removed to obtain the correct enter and merge selections on updates.